### PR TITLE
ZConduit

### LINKED ISSUE #2719: Port ZSink constructors and combinators to the new encoding
See current design in https://github.com/zio/zio/issues/2551#issuecomment-578415241.

Submit PRs to `feature/zconduit`. Reply below to claim an item.

Constructors:    
- [ ] ~await~ ported as `ZSink.head`
- [x] collectAll 
- [x] collectAllN - port as transducer
- [x] collectAllToMap 
- [x] collectAllToSet 
- [x] collectAllToMapN - port as transducer
- [x] collectAllToSetN - port as transducer
- [x] collectAllWhile - port as transducer
- [x] collectAllWhileM - port as transducer
- [x] die 
- [x] dieMessage 
- [x] drain 
- [x] fail
- [x] fold
- [x] foldLeft 
- [x] foldLeftM 
- [x] foldM
- [x] foldUntil - port as transducer
- [x] foldUntilM - port as transducer
- [x] foldWeighted - port as transducer
- [x] foldWeightedDecompose - port as transducer
- [x] foldWeightedDecomposeM - port as transducer
- [x] foldWeightedM - port as transducer 
- [x] fromEffect
- [x] fromFunction - will need a different signature (`A => Either[Z, Unit]`)
- [x] fromFunctionM - will need a different signature (`A => ZIO[R, Either[E, Z], Unit]`)
- [x] halt
- [x] head 
- [ ] ~identity~ - ported as `head`
- [x] ignoreWhile - port as transducer
- [x] ignoreWhileM - port as transducer
- [x] last 
- [ ] ~pull1~ requires leftovers to operate properly, don't port
- [ ] ~read1~ requires leftovers to operate properly, don't port
- [ ] splitOn - port as transducer
- [x] succeed
- [x] splitLines - port as transducer
- [ ] ~splitLinesChunk~ chunks are built-in so this will be ported as `splitLines`
- [x] utf8DecodeChunk - port as transducer

Combinators:
- [x] &> 
- [x] *> 
- [x] <& 
- [x] <&> 
- [x] <* 
- [x] <*> 
- [ ] <| 
- [x] as 
- [ ] asError 
- [x] contramap
- [x] contramapM
- [x] dimap
- [x] flatMap
- [x] map
- [x] mapError
- [x] mapM
- [ ] orElse
- [ ] provideSome
- [x] race 
- [x] raceBoth
- [ ] tapInput 
- [ ] tapOutput 
- [ ] timed
- [ ] unit 
- [x] zip 
- [x] zipLeft 
- [x] zipPar 
- [x] zipParLeft 
- [x] zipParRight 
- [x] zipRight 
- [x] zipWith 
- [x] zipWithPar
- [ ] ?
- [ ] chunked
- [ ] collectAll
- [ ] collectAllN
- [ ] collectAllWhile
- [x] collectAllWhileWith (signature changed to `Z => Boolean`
- [ ] collectAllWith
- [ ] optional
- [ ] takeWhile
- [x] untilOutput
- [ ] drop
- [ ] dropWhile
- [ ] filter
- [ ] filterM
- [ ] filterNot
- [ ] filterNotM
- [ ] keyed

### LINKED ISSUE #2720: Port ZStream combinators and constructors to the new encoding
Submit PRs to `feature/zconduit`. Reply below to claim an item.

Constructors:
- [x] absolve 
- [x] access 
- [x] accessM 
- [x] accessStream 
- [x] apply(A*)
- [x] bracket 
- [x] bracketExit 
- [x] crossN 
- [x] die 
- [x] dieMessage 
- [x] effectAsync 
- [x] effectAsyncInterrupt 
- [x] effectAsyncM 
- [x] effectAsyncMaybe 
- [x] environment 
- [x] fail 
- [x] finalizer 
- [x] flatten 
- [x] flattenPar 
- [x] flattenParUnbounded 
- [x] fromChunk 
- [x] fromEffect 
- [x] fromEffectOption 
- [x] fromInputStream 
- [x] fromIterable 
- [x] fromIterator 
- [x] fromIteratorManaged 
- [x] fromJavaIterator 
- [x] fromJavaIteratorManaged 
- [x] fromQueue 
- [x] fromQueueWithShutdown 
- [x] fromTQueue 
- [x] halt 
- [x] iterate 
- [x] managed 
- [x] mergeAll 
- [x] mergeAllUnbounded 
- [x] paginate 
- [x] paginateM 
- [x] range 
- [x] repeatEffect 
- [x] repeatEffectOption 
- [x] repeatEffectWith 
- [x] succeed 
- [x] unfold 
- [x] unfoldM 
- [x] unwrap 
- [x] unwrapManaged 
- [x] zipN 
- [x] empty 
- [x] never 
- [x] unit 

Combinators:
- [x] &> 
- [x] *> 
- [x] ++ 
- [x] <& 
- [x] <&> 
- [x] <* 
- [x] <*> 
- [x] absolve 
- [x] aggregate 
- [x] aggregateAsync
- [x] aggregateAsyncWithin 
- [x] aggregateAsyncWithinEither
- [ ] ~aggregateManaged~
- [x] as 
- [x] bimap 
- [x] broadcast 
- [x] broadcastDynamic 
- [x] broadcastedQueues
- [x] broadcastedQueuesDynamic
- [x] buffer 
- [x] bufferDropping 
- [x] bufferSignal 
- [x] bufferSliding 
- [x] bufferUnbounded 
- [x] catchAll 
- [x] catchAllCause
- [x] chunkN - port as transducer
- [x] collect 
- [x] collectM 
- [x] collectWhile 
- [x] collectWhileM 
- [x] combine 
- [x] concat 
- [x] cross 
- [x] crossWith 
- [x] distributedWith 
- [x] distributedWithDynamic 
- [x] drain 
- [x] drainFork 
- [x] drop 
- [x] dropUntil 
- [x] dropWhile 
- [x] either 
- [x] ensuring 
- [x] ensuringFirst 
- [x] filter 
- [x] filterM 
- [x] filterNot 
- [x] fixed 
- [x] flatMap
- [x] flatMapPar 
- [x] flatMapParSwitch 
- [x] fold 
- [x] foldM 
- [x] foldManaged 
- [x] foldManagedM 
- [x] foldWhile 
- [x] foldWhileM 
- [x] foldWhileManaged 
- [x] foldWhileManagedM 
- [x] foreach 
- [x] foreachManaged 
- [x] foreachWhile 
- [x] foreachWhileManaged 
- [x] forever 
- [x] groupBy
- [x] groupByKey 
- [x] grouped 
- [x] groupedWithin 
- [x] haltWhen 
- [x] interleave 
- [x] interleaveWith
- [x] interruptWhen 
- [x] into 
- [x] intoManaged 
- [x] map 
- [x] mapAccum 
- [x] mapAccumM 
- [x] mapConcat 
- [x] mapConcatChunk 
- [x] mapConcatChunkM 
- [x] mapConcatM 
- [x] mapError 
- [x] mapErrorCause 
- [x] mapM 
- [x] mapMPar 
- [x] mapMParUnordered 
- [x] mapMPartitioned 
- [x] merge 
- [x] mergeEither 
- [x] mergeWith
- [x] orElse 
- [x] partition 
- [x] partitionEither 
- [x] peel 
- [x] provide 
- [x] provideM 
- [x] provideManaged 
- [x] provideSome 
- [x] provideSomeM 
- [x] provideSomeManaged 
- [x] repeat 
- [x] repeatEither 
- [x] repeatWith 
- [x] run 
- [x] runCollect 
- [x] runDrain 
- [x] runHead 
- [x] runLast 
- [x] schedule 
- [x] scheduleEither 
- [x] scheduleElements 
- [x] scheduleElementsEither 
- [x] scheduleElementsWith 
- [x] scheduleWith 
- [x] take 
- [x] takeUntil 
- [x] takeWhile 
- [x] tap 
- [x] throttleEnforce (port as transducer)
- [x] throttleEnforceM (port as transducer)
- [x] throttleShape (port as transducer)
- [x] throttleShapeM (port as transducer)
- [x] timeout 
- [x] toInputStream 
- [x] toQueue 
- [x] toQueueUnbounded 
- [x] transduce 
- [ ] ~transduceManaged~
- [x] unNone 
- [x] unTake - became unExit and unExitChunk
- [x] via 
- [x] zip 
- [x] zipLeft 
- [x] zipRight 
- [x] zipWith
- [x] zipWithIndex 
- [x] zipWithLatest

### LINKED ISSUE #3347: The new ZStream/ZConduit design
Writing this up to document the current design for anyone looking to contribute.

### ZStream

The main change for `ZStream` is implicit chunking. This means that `ZStream` is now defined as:
```scala
class ZStream[R, E, O](pull: ZManaged[R, Nothing, ZIO[R, Option[E], Chunk[O]]])
```

Contract notes:
- Pulling a stream emits a chunk, an error (`ZIO.fail(Some(E))`) or an end-of-stream signal (`ZIO.fail(None)`).
- Pulling a stream after it has signaled end should keep failing with `None`.
- Pulling a stream after it has signaled an error should continue operating normally (not all combinators/constructors conform to this yet).
- There are no restrictions on the emptiness of chunks. Subsequently, combinators are free to pull upstream again until a non-empty chunk is emitted.

### ZSink

ZSink is now formulated as:
```scala
class ZSink[R, E, I, Z](push: ZManaged[R, Nothing, Option[Chunk[I]] => ZIO[R, Either[E, Z], Unit]])
```

Contract notes:
- A `Some(Chunk)` signals input from the stream; a `None` signals the end of the stream.
- Return values for sinks mean:
  `ZIO.unit` - "need more values"
  `ZIO.fail(Right(z))` - "ended with z"
  `ZIO.fail(Left(e))` - "failed with e"
- Sinks do *not* support remainders anymore. Supporting remainders in sinks proved to be difficult to do in a performant fashion, making them unattractive for parsing. Sinks should now be used mainly for composing functions of the form `Chunk[I] => ZIO[Unit]`.
- Sinks should consume entire chunks. If a sink consumes part of a chunk and drops the rest, it should probably be a transducer.
- Sinks should always end when receiving a `None`. It is a defect to not end with some sort of result (even a failure) when receiving a `None`.
- Sinks can assume they will not be pushed again after emitting a value.
- Sinks might be restarted as part of the stream's operation or by sink combinators (e.g. `ZSink#toTransducer, ZSink#collectAllWhileWith`).

### ZTransducer

Transducers are stateful stream functions that transform the stream's chunks. In contrast to sinks, they never end. Defined as:
```scala
class ZTransducer[R, E, I, O](push: ZManaged[R, Nothing, Option[Chunk[I]] => ZIO[R, E, Chunk[O]]])
```

Contract notes:
- A `Some(Chunk)` signals input from the stream; a `None` means "flush inner state".
- A transducer never ends and will not be restarted as part of the stream's operation. Therefore, it is safe to use the stateful layer to store remainders. For example, a transducer that groups elements into groups of 10 can store the leftovers from a chunk bigger than 10 elements in a Ref to be outputted on the next push.
- Transducers *must* flush any pending elements on receiving a `None`. It's not necessary to reset the internal state, but `push(None) *> push(None)` has to emit `Chunk.empty`.
- Transducers may emit empty chunks.

### ZConduit

Abstracting over all of these is `ZConduit`:
```scala
class ZConduit[R, E, I, O, Z](run: ZManaged[R, Nothing, Option[Chunk[I]] => ZIO[R, Either[E, Z], Chunk[O]]])
```

All the other constructs are specialized forms of ZConduit:
- `ZStream[R, E, O] = ZConduit[R, E, Unit, O, Unit]`
- `ZSink[R, E, I, Z] = ZConduit[R, E, I, Unit, Z]`
- `ZSink[R, E, I, O] = ZConduit[R, E, I, O, Nothing]`

It is still an open question (to me, at least 😄 ) whether ZConduit will be useful for users and implementors:
- For users, it could be interesting to treat streams/sinks/transducers as general conduits. However, I think that users usually want to know the specific type of conduit they're working with.
- For implementors, implementing operators on ZConduit can save multiple implementations across the constructs. However: 
  - it is usually more efficient to know the exact types of conduits you're composing (transducers compose differently with other transducers than with sinks)
  - it is useful to retain the exact type of the composed conduit, and that type depends on the conduits being composed (`transducer >>> transducer = transducer, transducer >>> sink = sink`, etc)
  - we'd like methods on `ZStream` to return `ZStream` and similarly for the other conduits. That means that if `ZStream <: ZConduit`, we need to override the methods inherited from ZConduit and refine the return types. That somewhat decreases the implementation benefit.

For now, we will implement most operations on the specific subtypes.

See also:
- #2551 for the original discussion
- #2719 for combinators/constructors that need to be ported from ZSink
- #2720 for combinators/constructors that need to be ported from ZStream (and some more discussions)

### LINKED ISSUE #3399: ZStream.fromInputStream blocking behavior and incorrect handling of 0 read bytes
It seems like the current implementation of `fromInputStream` is blocking and incorrectly handles the case where `bytesRead` is 0:
```scala
  def fromInputStream(
    is: => InputStream,
    chunkSize: Int = ZStreamChunk.DefaultChunkSize
  ): StreamEffectChunk[Any, IOException, Byte] =
    StreamEffectChunk {
      StreamEffect {
        Managed.effectTotal {
          var done       = false
          val capturedIs = is

          def pull(): Chunk[Byte] = {
            val buf = Array.ofDim[Byte](chunkSize)
            try {
              val bytesRead = capturedIs.read(buf) // <<<: blocks
              if (bytesRead < 0) {
                done = true
                end
              } else if (0 < bytesRead && bytesRead < buf.length) Chunk.fromArray(buf).take(bytesRead)
              else Chunk.fromArray(buf) // <<<: runs when bytesRead is 0 or buf.length
            } catch {
              case e: IOException => fail(e)
            }
          }

          () => if (done) end else pull()
        }
      }
    }
```

### PR DESCRIPTION
**Pull request by iravid**
Resolves #2719 
Resolves #2720 
Resolves #3347
Resolves #3399 
Resolves #3189

### CODE REVIEW

**Comment by iravid** [comment]
@regiskuckaertz I force-pushed the branch to my rebase yesterday so please rebase your open branch on it 🙏🏻

File: test/shared/src/main/scala/zio/test/Gen.scala [hunk 1]
```diff
@@ -151,11 +151,10 @@ final case class Gen[-R, +A](sample: ZStream[R, Nothing, Sample[R, A]]) { self =
   def zipWith[R1 <: R, B, C](that: Gen[R1, B])(f: (A, B) => C): Gen[R1, C] = Gen {
     val left  = self.sample.map(Right(_)) ++ self.sample.map(Left(_)).forever
     val right = that.sample.map(Right(_)) ++ that.sample.map(Left(_)).forever
-    left.zipWith(right) {
-      case (Some(Right(l)), Some(Right(r))) => Some(l.zipWith(r)(f))
-      case (Some(Right(l)), Some(Left(r)))  => Some(l.zipWith(r)(f))
-      case (Some(Left(l)), Some(Right(r)))  => Some(l.zipWith(r)(f))
-      case _                                => None
+    left.zipAllWith(right)(l => (Some(l), None), r => (None, Some(r)))((l, r) => (Some(l), Some(r))).collectWhile {
```
Comment (iravid) [inline test/shared/src/main/scala/zio/test/Gen.scala:154]: @adamgfraser Please double-check this

File: test/shared/src/main/scala/zio/test/Sample.scala [hunk 2]
```diff
@@ -98,28 +98,42 @@ final case class Sample[-R, +A](value: A, shrink: ZStream[R, Nothing, Sample[R,
     val shrink = self.shrink
       .combine[R1, Nothing, State, Sample[R1, B], Sample[R1, C]](that.shrink)((false, false, None, None)) {
         case ((leftDone, rightDone, s1, s2), left, right) =>
```
Comment (iravid) [inline test/shared/src/main/scala/zio/test/Sample.scala:100]: @adamgfraser And this

**Comment by adamgfraser** [comment]
@iravid Great!

**Pull request merged**